cpp/src/streaming/core/fanout.cpp

```diff
@@ -0,0 +1,297 @@
+ * @brief Broadcast messages from one input channel to multiple output channels.
+ *
+ * This is a general purpose implementation which can support consuming messages by any
+ * channel. A consumer node can decide to consume all messages from a single channel
+ * before moving to the next channel, or it can consume messages from all channels before
+ * moving to the next message. When a message has been sent to all output channels, it is
+ * purged from the internal deque.
+ *
+ * @param ctx The context to use.
+ * @param ch_in The input channel to receive messages from.
+ * @param chs_out The output channels to send messages to.
+ * @return A node representing the unbounded fanout operation.
+ */
+Node unbounded_fanout(
+    std::shared_ptr<Context> ctx,
+    std::shared_ptr<Channel> ch_in,
+    std::vector<std::shared_ptr<Channel>> chs_out
+) {
+    ShutdownAtExit ch_in_shutdown{ch_in};
+    ShutdownAtExit chs_out_shutdown{chs_out};
+    co_await ctx->executor()->schedule();
+
+    auto& logger = ctx->logger();
+
+    logger.debug("Scheduled unbounded fanout");
+    UnboundedFanoutState state(chs_out.size());
+
+    // start send tasks for each output channel
+    coro::task_container<coro::thread_pool> tasks(ctx->executor());
+    for (size_t i = 0; i < chs_out.size(); i++) {
+        RAPIDSMPF_EXPECTS(
+            tasks.start(unbounded_fo_send_task(*ctx, i, chs_out[i], state)),
+            "failed to start send task"
+        );
+    }
+
+    // input_done is only set by this task, so reading without lock is safe here
+    while (!state.input_done) {
+        {
+            auto lock = co_await state.mtx.scoped_lock();
+            co_await state.request_data.wait(lock, [&] {
+                return std::ranges::any_of(state.ch_next_idx, [&](size_t next_idx) {
+                    return state.recv_messages.size() == next_idx;
+                });
+            });
+        }
+
+        // receive a message from the input channel
+        auto msg = co_await ch_in->receive();
+
+        {  // relock mtx to update input_done/ recv_messages
+            auto lock = co_await state.mtx.scoped_lock();
+            if (msg.empty()) {
+                state.input_done = true;
+            } else {
+                logger.trace("Received input", msg.sequence_number());
+                state.recv_messages.emplace_back(std::move(msg));
+            }
+        }
+
+        // notify send_tasks to copy & send messages
+        co_await state.data_ready.notify_all();
+
+        // purge completed send_tasks. This will reset the messages to empty, so that they
+        // release the memory, however the deque is not resized. This guarantees that the
+        // indices are not invalidated.
+        // intentionally not locking the mtx here, because we only need to know a
+        // lower-bound on the last completed idx (ch_next_idx values are monotonically
+        // increasing)
+        size_t last_completed_idx = std::ranges::min(state.ch_next_idx);
```

Comment:
This is UB, because another thread (one of the consumers) might be updating an entry in `state.ch_next_idx` simultaneously.

Reply:
Yes, but its strictly increasing. And we only need an approximate value here. So, a current running `min` may not be the exact min in `std::ranges::min`, but it will be strictly less than or equal. Consequence would be, not cleaning up all the finished messages. But I felt it was a better trade-off than trying to relock the mutex.

Reply:
Hmm. Thinking about this again, maybe I can do a `ranges::min_max` during `request_data.wait`, and purge until the `min` value. That would eliminate this.